File: engine/simulation/engine.py
"""Discrete-time simulation engine — the heart of RouteX.

Each tick performs, in order:

1. advance the simulation clock
2. process scheduled events (road closure, accident, traffic spike, ...)
3. spawn due vehicles and move all active vehicles
4. update road occupancy, congestion and dynamic travel times
5. re-route vehicles whose roads changed state
6. update metrics

The engine is deterministic: the same scenario + seed + configuration always
produces the same behaviour.
"""

from __future__ import annotations

import random
from dataclasses import dataclass
from enum import StrEnum
from typing import Any

from engine.analytics.metrics import MetricsTracker
from engine.config import ScenarioConfig
from engine.network.edge import RoadStatus
from engine.network.network_builder import build_network
from engine.simulation.clock import SimulationClock
from engine.simulation.events import EventQueue, EventType, SimulationEvent
from engine.simulation.state import build_snapshot
from engine.traffic.traffic_lights import build_lights
from engine.vehicles.vehicle_manager import VehicleManager


class SimulationStatus(StrEnum):
    PENDING = "pending"
    RUNNING = "running"
    PAUSED = "paused"
    COMPLETED = "completed"
    STOPPED = "stopped"
    ERROR = "error"


@dataclass
class SimulationConfig:
    scenario_id: str = ""
    algorithm: str = "dijkstra"
    seed: int = 42
    max_ticks: int = 0
    speed: float = 1.0


class SimulationEngine:
    def __init__(
        self, scenario: ScenarioConfig, config: SimulationConfig | None = None
    ) -> None:
        self.scenario = scenario
        self.config = config or SimulationConfig(scenario_id=scenario.id)
        if self.config.max_ticks <= 0:
            self.config.max_ticks = scenario.duration
        self._rebuild()

    # ------------------------------------------------------------------ #
    # Lifecycle
    # ------------------------------------------------------------------ #
    def _rebuild(self) -> None:
        self.rng = random.Random(self.config.seed)
        self.clock = SimulationClock(tick_duration=1.0)
        self.network = build_network(self.scenario.network, self.rng)
        self.lights = build_lights(self.network, self.scenario.traffic_lights)
        self.manager = VehicleManager(
            self.network,
            self.scenario,
            self.rng,
            self.clock,
            self.lights,
            self.config.algorithm,
        )
        self.events = EventQueue()
        self.metrics = MetricsTracker()
        self.status = SimulationStatus.PENDING
        self.tick_events: list[dict[str, Any]] = []
        self._spike_ticks_left = 0
        self._spike_rate = 0
        self._schedule_events(self.scenario.events)

    def _schedule_events(self, raw_events: list[dict[str, Any]]) -> None:
        for raw in raw_events:
            self.events.schedule(SimulationEvent.from_dict(raw))

    def start(self) -> None:
        if self.status in (SimulationStatus.PENDING, SimulationStatus.PAUSED):
            self.status = SimulationStatus.RUNNING

    def pause(self) -> None:
        if self.status is SimulationStatus.RUNNING:
            self.status = SimulationStatus.PAUSED

    def stop(self) -> None:
        self.status = SimulationStatus.STOPPED

    def complete(self) -> None:
        if self.status is SimulationStatus.RUNNING:
            self.status = SimulationStatus.COMPLETED

    def reset(self) -> None:
        """Rebuild the simulation from scratch (deterministic replay)."""
        self._rebuild()

    # ------------------------------------------------------------------ #
    # Stepping
    # ------------------------------------------------------------------ #
    def step(self) -> None:
        if self.status is not SimulationStatus.RUNNING:
            return
        if self.clock.tick >= self.config.max_ticks:
            self.complete()
            return

        self.tick_events = []
        self.clock.step()
        self._process_events()
        self._process_traffic_spike()
        self.manager.step()
        self._refresh_network_state()
        self.metrics.update(self.clock.tick, self.clock.time, self.manager, self.network)

        if self.clock.tick >= self.config.max_ticks:
            self.complete()

    def run(self, steps: int) -> None:
        self.start()
        for _ in range(steps):
            if self.status is not SimulationStatus.RUNNING:
                break
            self.step()

    def snapshot(self) -> dict[str, Any]:
        return build_snapshot(self).to_dict()

    # ------------------------------------------------------------------ #
    # Internals
    # ------------------------------------------------------------------ #
    def _process_events(self) -> None:
        for event in self.events.pending(self.clock.tick):
            self.events.mark_applied(event)
            self.tick_events.append(event.to_dict())
            self._apply_event(event)

    def _apply_event(self, event: SimulationEvent) -> None:
        if event.type is EventType.ROAD_CLOSURE:
            self._set_road_status(event.road_id, RoadStatus.CLOSED)
        elif event.type is EventType.ROAD_REOPENING:
            self._set_road_status(event.road_id, RoadStatus.OPEN)
        elif event.type is EventType.ACCIDENT:
            self._set_road_status(event.road_id, RoadStatus.ACCIDENT)
        elif event.type is EventType.VEHICLE_SPAWN:
            self.manager.spawn_extra(int(event.payload.get("count", 1)))
        elif event.type is EventType.TRAFFIC_SPIKE:
            self._spike_rate = int(event.payload.get("count", 2))
            self._spike_ticks_left = max(1, event.duration)
        elif event.type is EventType.EMERGENCY_VEHICLE:
            self.manager.spawn_emergency(int(event.payload.get("count", 1)))
        elif event.type is EventType.TRAFFIC_LIGHT_CHANGE:
            self._change_light(event)

    def _set_road_status(self, road_id: str | None, status: RoadStatus) -> None:
        if road_id is None or road_id not in self.network.edges:
            return
        road = self.network.edges[road_id]
        road.status = status
        if status is not RoadStatus.OPEN:
            self.manager.reroute_vehicles_avoiding(road_id)

    def _change_light(self, event: SimulationEvent) -> None:
        if self.lights is None:
            return
        light = None
        if event.node_id:
            light = self.lights.light_for_node(event.node_id)
        if light is None and event.payload.get("light_id"):
            light = self.lights.lights.get(event.payload["light_id"])
        if light is None:
            return
        if "green_duration" in event.payload:
            light.green_duration = float(event.payload["green_duration"])
        if "red_duration" in event.payload:
            light.red_duration = float(event.payload["red_duration"])

    def _process_traffic_spike(self) -> None:
        if self._spike_ticks_left > 0:
            self.manager.spawn_extra(self._spike_rate)
            self._spike_ticks_left -= 1

    def _refresh_network_state(self) -> None:
        counts = {road_id: 0 for road_id in self.network.edges}
        for vehicle in self.manager.vehicles.values():
            if vehicle.current_edge in counts:
                counts[vehicle.current_edge] += 1
        for road_id, road in self.network.edges.items():
            road.current_vehicle_count = counts[road_id]
        self.network.refresh_dynamic_state()

File: browser-engine/engine/simulation/engine.ts
import { 
  ScenarioConfig, 
  ScenarioNetwork, 
  ScenarioTrafficLight, 
  RoadNetwork, 
  Road, 
  Vehicle, 
  VehicleType, 
  SimulationEvent, 
  EventType, 
  RoadStatus,
  SimulationStatus,
  SimulationConfig as SimConfig 
} from '../../shared/types';
import { SimulationClock } from './clock';
import { EventQueue, createEvent } from './events';
import { buildNetwork } from '../network/networkBuilder';
import { VehicleManager, createVehicleManager } from '../vehicles/vehicleManager';
import { refreshNetworkDynamicState } from '../traffic/trafficModel';
import { buildLights, stepLights, isGreen, getLightForNode } from '../traffic/trafficLights';
import { buildSnapshot, SimulationSnapshot } from './state';
import { createDijkstra, createAStar, defaultCost } from '../routing/algorithms';

export interface SimulationEngineConfig {
  scenario: ScenarioConfig;
  scenarioId?: string;
  algorithm?: string;
  seed?: number;
  maxTicks?: number;
  speed?: number;
}

export class SimulationEngine {
  private scenario: ScenarioConfig;
  private config: SimConfig;
  
  private clock!: SimulationClock;
  private network!: ReturnType<typeof buildNetwork>['network'];
  private lights!: Map<string, ReturnType<typeof buildLights> extends Map<string, infer T> ? T : never>;
  private manager!: VehicleManager;
  private events!: EventQueue;
  private status: SimulationStatus = SimulationStatus.PENDING;
  private tickEvents: SimulationEvent[] = [];
  
  private spikeTicksLeft = 0;
  private spikeRate = 0;
  
  private dijkstra = createDijkstra();
  private astar = createAStar();

  constructor(config: SimulationEngineConfig) {
    this.scenario = config.scenario;
    this.config = {
      scenarioId: config.scenarioId || config.scenario.id,
      algorithm: config.algorithm || 'dijkstra',
      seed: config.seed || 42,
      maxTicks: config.maxTicks || config.scenario.duration,
      speed: config.speed || 1.0,
    };
    this.rebuild();
  }

  rebuild(): void {
    const rng = this.createRNG(this.config.seed);
    
    this.clock = new SimulationClock(1.0);
    const { network } = buildNetwork(this.scenario.network);
    this.network = network;
    this.lights = buildLights(this.scenario.trafficLights);
    this.manager = createVehicleManager({
      network: this.network as RoadNetwork,
      rng,
      algorithm: this.config.algorithm,
    });
    this.manager.setSpawnRate(this.scenario.vehicleSpawnRate);
    this.events = new EventQueue();
    this.status = SimulationStatus.PENDING;
    this.tickEvents = [];
    this.spikeTicksLeft = 0;
    this.spikeRate = 0;
    this.scheduleEvents(this.scenario.events);
  }

  private createRNG(seed: number): () => number {
    let state = seed;
    return () => {
      state = (state * 1664525 + 1013904223) >>> 0;
      return state / 0x100000000;
    };
  }

  private scheduleEvents(rawEvents: Omit<SimulationEvent, 'id'>[]): void {
    for (const raw of rawEvents) {
      this.events.schedule(createEvent(
        raw.type,
        raw.tick,
        raw.duration,
        raw.roadId,
        raw.nodeId,
        raw.payload
      ));
    }
  }

  start(): void {
    if (this.status === SimulationStatus.PENDING || this.status === SimulationStatus.PAUSED) {
      this.status = SimulationStatus.RUNNING;
    }
  }

  pause(): void {
    if (this.status === SimulationStatus.RUNNING) {
      this.status = SimulationStatus.PAUSED;
    }
  }

  stop(): void {
    this.status = SimulationStatus.STOPPED;
  }

  step(): void {
    if (this.status !== SimulationStatus.RUNNING) return;
    if (this.clock.tick >= this.config.maxTicks) {
      this.complete();
      return;
    }

    this.tickEvents = [];
    this.clock.step();
    this.processEvents();
    this.processTrafficSpike();
    this.manager.step(this.clock.tick);
    this.updateNetworkState();
    this.stepTrafficLights();
    this.checkCompletion();
  }

  run(steps: number): void {
    this.start();
    for (let i = 0; i < steps; i++) {
      if (this.status !== SimulationStatus.RUNNING) break;
      this.step();
    }
  }

  snapshot(): SimulationSnapshot {
    return buildSnapshot(this);
  }

  getTick(): number {
    return this.clock.tick;
  }

  getTime(): number {
    return this.clock.time;
  }

  getStatus(): SimulationStatus {
    return this.status;
  }

  getVehicles(): Vehicle[] {
    return this.manager.vehiclesList;
  }

  getNetwork(): RoadNetwork {
    return this.network as RoadNetwork;
  }

  getNetworkSummary(): { nodes: number; edges: number; closed: number } {
    let closed = 0;
    for (const road of this.network.edges.values()) {
      if (road.status !== RoadStatus.OPEN) closed++;
    }
    return { nodes: this.network.nodes.size, edges: this.network.edges.size, closed };
  }

  getAverageCongestion(): number {
    let total = 0;
    let count = 0;
    for (const road of this.network.edges.values()) {
      total += road.congestion;
      count++;
    }
    return count > 0 ? total / count : 0;
  }

  getTotalWaitingTime(): number {
    let total = 0;
    for (const vehicle of this.manager.vehiclesList) {
      if (!vehicle.arrived && vehicle.currentEdge) {
        const road = this.network.edges.get(vehicle.currentEdge);
        if (road && road.congestion > 0.5) {
          total += 1;
        }
      }
    }
    return total;
  }

  getEmergencyResponseTime(): number | null {
    const emergencies = this.manager.vehiclesList.filter(v => v.type === VehicleType.EMERGENCY && v.arrived);
    if (emergencies.length === 0) return null;
    let total = 0;
    for (const v of emergencies) {
      if (v.arrivalTick !== null) {
        total += v.arrivalTick - v.spawnTick;
      }
    }
    return total / emergencies.length;
  }

  private processEvents(): void {
    for (const event of this.events.pending(this.clock.tick)) {
      this.events.markApplied(event);
      this.tickEvents.push(event);
      this.applyEvent(event);
    }
  }

  private applyEvent(event: SimulationEvent): void {
    switch (event.type) {
      case EventType.ROAD_CLOSURE:
        this.setRoadStatus(event.roadId, RoadStatus.CLOSED);
        break;
      case EventType.ROAD_REOPENING:
        this.setRoadStatus(event.roadId, RoadStatus.OPEN);
        break;
      case EventType.ACCIDENT:
        this.setRoadStatus(event.roadId, RoadStatus.ACCIDENT);
        break;
      case EventType.VEHICLE_SPAWN:
        this.manager.spawnExtra(event.payload.count as number || 1);
        break;
      case EventType.TRAFFIC_SPIKE:
        this.spikeRate = (event.payload.count as number) || 2;
        this.spikeTicksLeft = Math.max(1, event.duration);
        break;
      case EventType.EMERGENCY_VEHICLE:
        this.manager.spawnEmergency(event.payload.count as number || 1);
        break;
      case EventType.TRAFFIC_LIGHT_CHANGE:
        this.changeLight(event);
        break;
    }
  }

  private setRoadStatus(roadId: string | null, status: RoadStatus): void {
    if (!roadId || !this.network.edges.has(roadId)) return;
    const road = this.network.edges.get(roadId)!;
    road.status = status;
    if (status !== RoadStatus.OPEN) {
      this.manager.rerouteVehiclesAvoiding(roadId);
    }
  }

  private changeLight(event: SimulationEvent): void {
    const light = event.nodeId 
      ? getLightForNode(this.lights, event.nodeId)
      : this.lights.get(event.payload.lightId as string);
    if (!light) return;
    if ('green_duration' in event.payload) {
      light.greenDuration = event.payload.green_duration as number;
    }
    if ('red_duration' in event.payload) {
      light.redDuration = event.payload.red_duration as number;
    }
  }

  private processTrafficSpike(): void {
    if (this.spikeTicksLeft > 0) {
      this.manager.spawnExtra(this.spikeRate);
      this.spikeTicksLeft--;
    }
  }

  private updateNetworkState(): void {
    const vehicleCounts = new Map<string, number>();
    for (const vehicle of this.manager.vehiclesList) {
      if (vehicle.currentEdge) {
        vehicleCounts.set(vehicle.currentEdge, (vehicleCounts.get(vehicle.currentEdge) || 0) + 1);
      }
    }
    refreshNetworkDynamicState(this.network.edges, vehicleCounts);
  }

  private stepTrafficLights(): void {
    stepLights(this.lights);
  }

  private checkCompletion(): void {
    if (this.clock.tick >= this.config.maxTicks) {
      this.complete();
    }
  }

  private complete(): void {
    if (this.status === SimulationStatus.RUNNING) {
      this.status = SimulationStatus.COMPLETED;
    }
  }
}

export function createSimulationEngine(config: SimulationEngineConfig): SimulationEngine {
  return new SimulationEngine(config);
}